### PR TITLE
Update and extend the file tests and fix our file support

### PR DESCRIPTION
**Pull request by robgjansen**
Converts file test to use test macros. Includes better testing of temp files and mmapping unlinked files. Fixes a directory bug (found while running Tor).

### CODE REVIEW

File: src/main/host/descriptor/file.c [hunk 1]
```diff
@@ -194,87 +183,118 @@ static char* _file_getPath(File* file, File* dir, const char* pathname) {
...
-    }
-    return abspath;
-}
+#ifdef DEBUG
+#define CHECK_FLAG(flag)                                                                           \
+    if (flags & flag) {                                                                            \
+        if (!flag_str) {                                                                           \
+            asprintf(&flag_str, #flag);                                                            \
+        } else {                                                                                   \
+            flag_str = _file_getConcatStr(flag_str, '|', #flag);                                   \
```
Comment (sporksmith) [inline src/main/host/descriptor/file.c:197]: doesn't this leak the old `flag_str` buffer?

File: src/main/host/descriptor/file.c [hunk 2]
```diff
@@ -194,87 +183,118 @@ static char* _file_getPath(File* file, File* dir, const char* pathname) {
...
+    }
+#endif
 
-    debug("Attempting to open file with pathname '%s'", pathname);
+    /* The default case is a regular file. We do this first so that we have
+     * an absolute path to compare for special files. */
+    char* abspath = _file_getPath(file, dir, pathname);
 
+    // When used with O_TMPFILE, O_EXCL will prevent the temporary file from later being linked
+    // into the filesystem. We unset O_EXCL since we may need that functionality, e.g. if the
```
Comment (sporksmith) [inline src/main/host/descriptor/file.c:247]: Do we still need this with the "open `/proc/<shadow pid>/..../<fd>`" strategy?

File: src/main/host/descriptor/file.c [hunk 3]
```diff
@@ -194,87 +183,118 @@ static char* _file_getPath(File* file, File* dir, const char* pathname) {
...
         if (abspath) {
             free(abspath);
         }
         file->type = FILE_TYPE_NOTSET;
-        return -errno;
+        return -errcode;
     }
 
-    /* Store the create information so we can open later if needed. */
+    /* Store the create information so we can re-open later if needed (e.g., mmap). */
```
Comment (sporksmith) [inline src/main/host/descriptor/file.c:294]: Still needed?

File: src/shim/preload_libraries.c [hunk 4]
```diff
@@ -457,44 +457,6 @@ int gethostname(char* name, size_t len) {
     return 0;
 }
 
-static void _delete_file_on_exit(int status, void* filename) {
```
Comment (sporksmith) [inline src/shim/preload_libraries.c:460]: Woohoo!

**Review by sporksmith — CHANGES_REQUESTED** [review]

File: src/main/host/descriptor/file.c (hunk 1 above)
Comment (robgjansen) [reply src/main/host/descriptor/file.c:197]: Oops, good catch!

File: src/main/host/descriptor/file.c (hunk 3 above)
Comment (robgjansen) [reply src/main/host/descriptor/file.c]: We use `osfd`, `flags`, and `mode` when we `mmap` the file, and `abspath` for logging.

**Review by sporksmith — APPROVED** [review]

**Pull request merged**